File: src/main/java/io/github/ytg1234/songsounds/command/SetsongCommand.java
package io.github.ytg1234.songsounds.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.suggestion.SuggestionProvider;
import io.github.ytg1234.songsounds.SongManager;
import io.github.ytg1234.songsounds.base.song.Song;
import io.github.ytg1234.songsounds.util.ModVars;
import net.minecraft.command.CommandSource;
import net.minecraft.command.argument.IdentifierArgumentType;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.text.TranslatableText;

import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class SetsongCommand {
    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        final SuggestionProvider<ServerCommandSource>
                SUGGESTION_PROVIDER =
                (commandContext, suggestionsBuilder) -> CommandSource.suggestIdentifiers(SongManager.INSTANCE.getSongIDs(), suggestionsBuilder);

        dispatcher.register(literal("setsong").then(argument("id", IdentifierArgumentType.identifier()).suggests(SUGGESTION_PROVIDER)
                                                                                                       .executes(SetsongCommand::execute)));
    }

    private static int execute(CommandContext<ServerCommandSource> ctx) {
        Song song = SongManager.INSTANCE.getSong(IdentifierArgumentType.getIdentifier(ctx, "id"));

        if (song == null || song.equals(Song.EMPTY) || IdentifierArgumentType.getIdentifier(ctx, "id").equals(Song.EMPTY_ID)) {
            ctx.getSource().sendError(new TranslatableText("error.songsounds.emptysong"));
            return 0;
        } else if (song.sections == null || song.sections.length <= 0) {
            ctx.getSource().sendError(new TranslatableText("error.songsounds.nosections"));
            return 0;
        } else if (song.sections[0].notes == null || song.sections[0].notes.length <= 0) {
            ctx.getSource().sendError(new TranslatableText("error.songsounds.nonotes"));
            return 0;
        } else if (song.equals(ModVars.currentSong)) {
            ctx.getSource().sendError(new TranslatableText("error.songsounds.same", song.name));
            return 0;
        } else if (song.name == null) {
            ctx.getSource().sendError(new TranslatableText("error.songsounds.other"));
            return 0;
        }

        ModVars.currentSong = song;
        ModVars.index = 0;
        ModVars.section = 0;

        ctx.getSource().sendFeedback(new TranslatableText("text.songsounds.switched", song.name), true);
        return 1;
    }
}

File: src/main/kotlin/io/github/ytg1234/songsounds/command/SetsongCommand.kt
package io.github.ytg1234.songsounds.command

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.suggestion.SuggestionProvider
import io.github.ytg1234.songsounds.base.SongManager
import io.github.ytg1234.songsounds.base.song.Song
import io.github.ytg1234.songsounds.util.Builder
import io.github.ytg1234.songsounds.util.PermedCommand
import io.github.ytg1234.songsounds.util.currentSong
import io.github.ytg1234.songsounds.util.setSongIdAndReset
import net.minecraft.command.CommandSource
import net.minecraft.command.argument.IdentifierArgumentType
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.TranslatableText

object SetsongCommand : PermedCommand("setsong", "songsounds.command.setsong", 2) {
    override val cmd = { it: Builder ->
        val provider = SuggestionProvider<ServerCommandSource> { _, builder ->
            CommandSource.suggestIdentifiers(
                SongManager.songIDs,
                builder
            )
        }

        it.then(
            CommandManager.argument("id", IdentifierArgumentType.identifier()).suggests(provider).executes(::execute)
        )
    }

    private fun execute(ctx: CommandContext<ServerCommandSource>): Int {
        val id = IdentifierArgumentType.getIdentifier(ctx, "id")
        val song: Song = SongManager[id]

        if (song == Song.EMPTY || IdentifierArgumentType.getIdentifier(ctx, "id") == Song.EMPTY_ID) {
            ctx.source.sendError(TranslatableText("error.songsounds.emptysong"))
            return 0
        } else if (song.sections.isEmpty()) {
            ctx.source.sendError(TranslatableText("error.songsounds.nosections"))
            return 0
        } else if (song.sections[0].notes.isEmpty()) {
            ctx.source.sendError(TranslatableText("error.songsounds.nonotes"))
            return 0
        } else if (song == currentSong) {
            ctx.source.sendError(TranslatableText("error.songsounds.same", song.name))
            return 0
        } else if (song.name == null) {
            ctx.source.sendError(TranslatableText("error.songsounds.other"))
            return 0
        }

        setSongIdAndReset(IdentifierArgumentType.getIdentifier(ctx, "id"))
        ctx.source.sendFeedback(TranslatableText("text.songsounds.switched", song.name), true)

        return Command.SINGLE_SUCCESS
    }
}
